packages/controls/src/controls/shape/v2/shape-v2.ts

```diff
@@ -214,28 +209,50 @@ class Definition<C extends Config> extends ControlDefinition<
   ): Data {
     if (translatedData == null) return data
 
-    const versionedData = Definition.versionData(data)
+    const itemsData = Definition.itemsData(data)
 
     return mapValues(this.keyDefs, (def, key) =>
-      def.mergeTranslatedData(
-        versionedData.value[key],
-        translatedData[key],
-        context,
-      ),
+      def.mergeTranslatedData(itemsData[key], translatedData[key], context),
+    )
+  }
+
+  static isShapeV2Data<C extends Config>(
+    data: DataType<C>,
+  ): data is ShapeV2DataTypeV1<C> {
+    return (
+      data != null &&
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === this.v1DataType
     )
   }
 
-  static versionData<C extends Config>(
+  static itemsData<C extends Config>(
+    data: DataType<C>,
+  ): ShapeV2DataTypeV1<C>['value'] {
+    if (Definition.isShapeV2Data(data)) {
+      return data.value
+    }
+
+    return data
+  }
+
+  static itemDataPath<C extends Config>(
+    data: DataType<C>,
+    key: string,
+  ): string[] {
+    return Definition.isShapeV2Data(data) ? ['value', key] : [key]
+  }
```

Comment:
Needed by the builder, implemented as a part of the `ShapeV2` interface to keep the knowledge about the inner structure of the shape data in a single place.